File: Source/CSharp/Analyzers/ReservedNameAnalyzer.cs
// The MIT License (MIT)
// 
// Copyright (c) 2014-2015, Institute for Software & Systems Engineering
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

namespace SafetySharp.CSharp.Analyzers
{
	using System;
	using System.Linq;
	using Microsoft.CodeAnalysis;
	using Microsoft.CodeAnalysis.CSharp;
	using Microsoft.CodeAnalysis.CSharp.Syntax;
	using Microsoft.CodeAnalysis.Diagnostics;
	using Roslyn;
	using Roslyn.Syntax;
	using Utilities;

	/// <summary>
	///     Ensures that no type, type member, variable, etc. uses a name reserved for synthesized variables.
	/// </summary>
	[DiagnosticAnalyzer, UsedImplicitly]
	public class ReservedNameAnalyzer : CSharpAnalyzer
	{
		/// <summary>
		///     The error diagnostic emitted by the analyzer.
		/// </summary>
		private static readonly DiagnosticInfo ReservedName = DiagnosticInfo.Error(
			DiagnosticIdentifier.ReservedName,
			"The identifier name is reserved for internal use.",
			"Identifier name '{0}' is reserved for internal use.");

		/// <summary>
		///     Initializes a new instance.
		/// </summary>
		public ReservedNameAnalyzer()
			: base(ReservedName)
		{
		}

		/// <summary>
		///     Called once at session start to register actions in the analysis context.
		/// </summary>
		/// <param name="context">The analysis context that should be used to register analysis actions.</param>
		public override void Initialize(AnalysisContext context)
		{
			context.RegisterSyntaxTreeAction(Analyze);
		}

		/// <summary>
		///     Performs the analysis.
		/// </summary>
		/// <param name="context">The context in which the analysis should be performed.</param>
		private static void Analyze(SyntaxTreeAnalysisContext context)
		{
			var syntaxTree = context.Tree;

			CheckIdentifiers<BaseTypeDeclarationSyntax>(context, syntaxTree, t => t.Identifier);
			CheckIdentifiers<EnumMemberDeclarationSyntax>(context, syntaxTree, e => e.Identifier);
			CheckIdentifiers<ParameterSyntax>(context, syntaxTree, p => p.Identifier);
			CheckIdentifiers<VariableDeclaratorSyntax>(context, syntaxTree, v => v.Identifier);
			CheckIdentifiers<MethodDeclarationSyntax>(context, syntaxTree, m => m.Identifier);
			CheckIdentifiers<EventDeclarationSyntax>(context, syntaxTree, e => e.Identifier);
			CheckIdentifiers<PropertyDeclarationSyntax>(context, syntaxTree, p => p.Identifier);

			var invalidNamespaces = syntaxTree
				.Descendants<NamespaceDeclarationSyntax>()
				.SelectMany(n => n.Descendants<IdentifierNameSyntax>())
				.Where(identifierName => identifierName.Identifier.IsSynthesized());

			foreach (var invalidNamespace in invalidNamespaces)
				ReservedName.Emit(context, invalidNamespace.Identifier, invalidNamespace.Identifier.ValueText);
		}

		/// <summary>
		///     Checks the identifier of all nodes of type <typeparamref name="T" /> within <paramref name="syntaxTree" />.
		///     The <paramref name="getIdentifier" /> function is used to extract the identifier
		///     <see cref="SyntaxToken" /> from the <see cref="SyntaxNode" />.
		/// </summary>
		/// <typeparam name="T">The type of the <see cref="SyntaxNode" />s that should be checked.</typeparam>
		/// <param name="context">The context in which the analysis should be performed.</param>
		/// <param name="syntaxTree">The <see cref="SyntaxTree" /> that should be checked.</param>
		/// <param name="getIdentifier">
		///     A function that gets the <see cref="SyntaxToken" /> representing the identifier of a
		///     <see cref="SyntaxNode" /> of type <typeparamref name="T" />.
		/// </param>
		private static void CheckIdentifiers<T>(SyntaxTreeAnalysisContext context, SyntaxTree syntaxTree, Func<T, SyntaxToken> getIdentifier)
			where T : CSharpSyntaxNode
		{
			var invalidIdentifiers = syntaxTree
				.Descendants<T>()
				.Select(getIdentifier)
				.Where(name => name.IsSynthesized());

			foreach (var identifier in invalidIdentifiers)
				ReservedName.Emit(context, identifier, identifier.ValueText);
		}
	}
}

File: Tests/SafetySharp/Analyzers/ReservedNameAnalyzer.fs
// The MIT License (MIT)
// 
// Copyright (c) 2014-2015, Institute for Software & Systems Engineering
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

namespace Analyzers

open System
open System.Linq
open NUnit.Framework
open SafetySharp.Modeling
open SafetySharp.Compiler.Analyzers
open SafetySharp.Compiler.Roslyn
open SafetySharp.Compiler.Roslyn.Syntax
open SafetySharp.Compiler.Roslyn.Symbols

[<TestFixture>]
module ``Rreserved names`` =
    let getDiagnostic = TestCompilation.GetDiagnostic (ReservedNameAnalyzer ())

    let diagnostic location identifierName =
        errorDiagnostic DiagnosticIdentifier.ReservedName (1, location) (1, location + String.length identifierName)
            "Identifier name '%s' is reserved for internal use." identifierName

    [<Test>]
    let ``non-reserved names are valid`` () =
        getDiagnostic "enum E { A }" =? None
        getDiagnostic "interface I { void M(int a); int P { get; set; } }" =? None
        getDiagnostic "class C { void M(int a) { int b = a; } int P { get; set; } event System.EventHandler A; }" =? None
        getDiagnostic "struct C { void M(int a) { int b = a; } int P { get; set; } }" =? None
        getDiagnostic "namespace N {}" =? None
        getDiagnostic "namespace N.M {}" =? None

    [<Test>]
    let ``reserved names are invalid`` () =
        let invalidName = "Name".ToSynthesized ()
        getDiagnostic (sprintf "enum %s : byte { A }" invalidName) =? diagnostic 5 invalidName
        getDiagnostic (sprintf "enum E : byte { %s }"  invalidName)=? diagnostic 16 invalidName
        getDiagnostic (sprintf "interface %s { void M(int a); int P { get; set; } }"  invalidName)=? diagnostic 10 invalidName
        getDiagnostic (sprintf "interface I { void %s(int a); int P { get; set; } }"  invalidName)=? diagnostic 19 invalidName
        getDiagnostic (sprintf "interface I { void M(int %s); int P { get; set; } }"  invalidName)=? diagnostic 25 invalidName
        getDiagnostic (sprintf "interface I { void M(int a); int %s { get; set; } }"  invalidName)=? diagnostic 33 invalidName
        getDiagnostic (sprintf "class %s { void M(int a) { int b = a; } int P { get; set; } }" invalidName) =? diagnostic 6 invalidName
        getDiagnostic (sprintf "class C { void %s(int a) { int b = a; } int P { get; set; } }" invalidName) =? diagnostic 15 invalidName
        getDiagnostic (sprintf "class C { void M(int %s) { int b = 1; } int P { get; set; } }" invalidName) =? diagnostic 21 invalidName
        getDiagnostic (sprintf "class C { void M(int a) { int %s = a; } int P { get; set; } }" invalidName) =? diagnostic 30 invalidName
        getDiagnostic (sprintf "class C { void M(int a) { int %s = a, x; } int P { get; set; } }" invalidName) =? diagnostic 30 invalidName
        getDiagnostic (sprintf "class C { void M(int a) { int x, %s = a; } int P { get; set; } }" invalidName) =? diagnostic 33 invalidName
        getDiagnostic (sprintf "class C { void M(int a) { int b = a; } int %s { get; set; } }" invalidName) =? diagnostic 43 invalidName
        getDiagnostic (sprintf "class C { int %s; }" invalidName) =? diagnostic 14 invalidName
        getDiagnostic (sprintf "class C { int %s, a; }" invalidName) =? diagnostic 14 invalidName
        getDiagnostic (sprintf "class C { int a, %s; }" invalidName) =? diagnostic 17 invalidName
        getDiagnostic (sprintf "class C { event System.EventHandler %s; }" invalidName) =? diagnostic 36 invalidName
        getDiagnostic (sprintf "class C { event System.EventHandler %s, a; }" invalidName) =? diagnostic 36 invalidName
        getDiagnostic (sprintf "class C { event System.EventHandler a, %s; }" invalidName) =? diagnostic 39 invalidName
        getDiagnostic (sprintf "class C { event System.EventHandler %s { add {} remove {} } }" invalidName) =? diagnostic 36 invalidName
        getDiagnostic (sprintf "struct %s { void M(int a) { int b = a; } int P { get; set; } }" invalidName) =? diagnostic 7 invalidName
        getDiagnostic (sprintf "struct S { void %s(int a) { int b = a; } int P { get; set; } }" invalidName) =? diagnostic 16 invalidName
        getDiagnostic (sprintf "struct S { void M(int %s) { int b = 1; } int P { get; set; } }" invalidName) =? diagnostic 22 invalidName
        getDiagnostic (sprintf "struct S { void M(int a) { int %s = a; } int P { get; set; } }" invalidName) =? diagnostic 31 invalidName
        getDiagnostic (sprintf "struct S { void M(int a) { int %s = a, x; } int P { get; set; } }" invalidName) =? diagnostic 31 invalidName
        getDiagnostic (sprintf "struct S { void M(int a) { int x, %s = a; } int P { get; set; } }" invalidName) =? diagnostic 34 invalidName
        getDiagnostic (sprintf "struct S { void M(int a) { int b = a; } int %s { get; set; } }" invalidName) =? diagnostic 44 invalidName
        getDiagnostic (sprintf "struct S { int %s; }" invalidName) =? diagnostic 15 invalidName
        getDiagnostic (sprintf "struct S { int %s, a; }" invalidName) =? diagnostic 15 invalidName
        getDiagnostic (sprintf "struct S { int a, %s; }" invalidName) =? diagnostic 18 invalidName
        getDiagnostic (sprintf "struct S { event System.EventHandler %s; }" invalidName) =? diagnostic 37 invalidName
        getDiagnostic (sprintf "struct S { event System.EventHandler %s, a; }" invalidName) =? diagnostic 37 invalidName
        getDiagnostic (sprintf "struct S { event System.EventHandler a, %s; }" invalidName) =? diagnostic 40 invalidName
        getDiagnostic (sprintf "struct S { event System.EventHandler %s { add {} remove {} } }" invalidName) =? diagnostic 37 invalidName
        getDiagnostic (sprintf "namespace %s {}" invalidName) =? diagnostic 10 invalidName
        getDiagnostic (sprintf "namespace %s.M {}" invalidName) =? diagnostic 10 invalidName
        getDiagnostic (sprintf "namespace N.%s {}" invalidName) =? diagnostic 12 invalidName

    [<Test>]
    let ``reserved names within expressions are valid`` () =
        // Only emits a diagnostic for the declaration of the method parameter, but not for the usages of the parameter.
        let invalidName = "Name".ToSynthesized ()
        getDiagnostic (sprintf "class X { void M(int %s) { int b = %s + %s; } }" invalidName invalidName invalidName) =? diagnostic 21 invalidName
        getDiagnostic (sprintf "class %s { void M() { var t = typeof(%s); } }" invalidName invalidName) =? diagnostic 6 invalidName
